crates/matrix-sdk-crypto/src/session_manager/sessions.rs

```diff
@@ -257,43 +260,70 @@ impl SessionManager {
                             .or_default()
-                            .insert(device_id, DeviceKeyAlgorithm::SignedCurve25519);
+                            .insert(device_id);
                     }
                 } else {
-                    warn!(
-                        user_id = ?device.user_id(),
-                        device_id = ?device.device_id(),
-                        "Device doesn't have a valid Curve25519 key, \
-                        can't establish an Olm session"
-                    );
+                    bad_key_devices.entry(user_id.clone()).or_default().insert(device_id);
                 }
             }
         }
 
         // Add the list of sessions that for some reason automatically need to
         // create an Olm session.
         for (user, device_ids) in self.users_for_key_claim.read().unwrap().iter() {
-            for device_id in device_ids {
-                missing
-                    .entry(user.to_owned())
-                    .or_default()
-                    .insert(device_id.to_owned(), DeviceKeyAlgorithm::SignedCurve25519);
-            }
+            missing_session_devices_by_user
+                .entry(user.to_owned())
+                .or_default()
+                .extend(device_ids.iter().cloned());
         }
 
-        let result = if missing.is_empty() {
+        debug!(
+            ?missing_session_devices_by_user,
+            ?timed_out_devices_by_user,
+            "Collected user/device pairs that are missing an Olm session"
+        );
+
+        if !non_olm_devices_by_user.is_empty() {
+            warn!(
+                ?non_olm_devices_by_user,
+                "Some devices don't support any of our 1-to-1 E2EE algorithms: \
+                can't establish an Olm session with them"
+            );
+        }
+
+        if !bad_key_devices.is_empty() {
+            warn!(
+                ?bad_key_devices,
+                "Some devices don't have a valid Curve25519 key: \
+                can't establish an Olm session with them"
+            );
+        }
+
+        let result = if missing_session_devices_by_user.is_empty() {
             None
         } else {
-            debug!(
-                ?missing,
-                ?timed_out,
-                "Collected user/device pairs that are missing an Olm session"
-            );
+            // rewrite the missing_session_devices_by_user map to include the desired
+            // algorithm for each device
+            let missing: BTreeMap<_, _> = missing_session_devices_by_user
+                .into_iter()
+                .map(|(user_id, user_devices)| {
+                    (
+                        user_id,
+                        user_devices
+                            .into_iter()
+                            .map(|device_id| (device_id, DeviceKeyAlgorithm::SignedCurve25519))
+                            .collect::<BTreeMap<_, _>>(),
+                    )
+                })
+                .collect();
```

Comment:
since we no longer include the algorithm when building `missing_session_devices_by_user`, we now have to rebuild the whole structure to have the right shape to pass into `KeysClaimRequest::new`.